kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,442 @@
+                    Error::change_data_feed_incompatible_schema(table_schema, &schema)
+                );
+                let table_properties = TableProperties::from(configuration);
+                check_cdf_table_properties(&table_properties)
+                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
+            }
+        }
+        // We resolve the remove deletion vector map after visiting the entire commit.
+        if has_cdc_action {
+            remove_dvs.clear();
+        } else {
+            // The only (path, deletion_vector) pairs we must track are ones whose path is the
+            // same as an `add` action.
+            remove_dvs.retain(|rm_path, _| add_paths.contains(rm_path));
+        }
+        Ok(LogReplayScanner {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            has_cdc_action,
+            remove_dvs,
+        })
+    }
+    /// Generates an iterator of [`TableChangesScanData`] that iterates over each action of the
+    /// commit. This performs phase 2 of [`LogReplayScanner`].
+    fn into_scan_batches(
+        self,
+        engine: Arc<dyn Engine>,
+        filter: Option<Arc<DataSkippingFilter>>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+        let Self {
+            has_cdc_action,
+            remove_dvs,
+            commit_file,
+            // TODO: Add the timestamp as a column with an expression
+            timestamp: _,
+        } = self;
+        let remove_dvs = Arc::new(remove_dvs);
+
+        let schema = FileActionSelectionVisitor::schema()?;
+        let action_iter = engine.get_json_handler().read_json_files(
+            &[commit_file.location.clone()],
+            schema,
+            None,
+        )?;
+
+        let result = action_iter.map(move |actions| -> DeltaResult<_> {
+            let actions = actions?;
+
+            // Apply data skipping to get back a selection vector for actions that passed skipping.
+            // We start our selection vector based on what was filtered. We will add to this vector
+            // below if a file has been removed. Note: None implies all files passed data skipping.
+            let selection_vector = match &filter {
+                Some(filter) => filter.apply(actions.as_ref())?,
+                None => vec![true; actions.len()],
+            };
+
+            let mut visitor =
+                FileActionSelectionVisitor::new(&remove_dvs, selection_vector, has_cdc_action);
+            visitor.visit_rows_of(actions.as_ref())?;
+            let data = engine
+                .get_expression_handler()
+                .get_evaluator(
+                    get_log_add_schema().clone(),
+                    get_add_transform_expr(),
+                    scan_row_schema().into(),
+                )
+                .evaluate(actions.as_ref())?;
+            Ok(TableChangesScanData {
+                data,
```

Comment:
nit if we do the renames above
```suggestion
            let scan_data = engine
                .get_expression_handler()
                .get_evaluator(
                    get_log_add_schema().clone(),
                    get_add_transform_expr(),
                    scan_data_schema().into(),
                )
                .evaluate(actions.as_ref())?;
            Ok(TableChangesScanData {
                scan_data,
```